src/libs/relay/conduit_relay_mpi.cpp

```diff
@@ -113,171 +113,618 @@ rank(MPI_Comm mpi_comm)
+        case CONDUIT_NATIVE_DOUBLE_ID: res = MPI_DOUBLE; break;
+
+        #if defined(CONDUIT_USE_LONG_DOUBLE)
+        case CONDUIT_NATIVE_LONG_DOUBLE_ID: res = MPI_LONG_DOUBLE; break;
+        #endif
+
+
+        case DataType::CHAR8_STR_ID:  res = MPI_CHAR;  break;
+
+    }
+    
+    return res;
+}
+
+
+//-----------------------------------------------------------------------------
+index_t
+mpi_dtype_to_conduit_dtype_id(MPI_Datatype dt)
+{
+    index_t res = DataType::EMPTY_ID;
+
+    switch(dt)
+    {
+        // string type
+        case MPI_CHAR:  res = DataType::CHAR8_STR_ID;  break;
+
+        // signed integer types
+        case MPI_SHORT: res = CONDUIT_NATIVE_SHORT_ID; break;
+        case MPI_INT:   res = CONDUIT_NATIVE_INT_ID;   break;
+        case MPI_LONG:  res = CONDUIT_NATIVE_LONG_ID;  break;
+
+        #if defined(CONDUIT_USE_LONG_LONG)
+        case MPI_LONG_LONG: res = CONDUIT_NATIVE_LONG_LONG_ID; break;
+        #endif
+
+        // unsigned integer types 
+        case MPI_BYTE:           res = CONDUIT_NATIVE_UNSIGNED_CHAR_ID;  break;
+        case MPI_UNSIGNED_CHAR:  res = CONDUIT_NATIVE_UNSIGNED_CHAR_ID;  break;
+        case MPI_UNSIGNED_SHORT: res = CONDUIT_NATIVE_UNSIGNED_SHORT_ID; break;
+        case MPI_UNSIGNED:       res = CONDUIT_NATIVE_UNSIGNED_INT_ID;   break;
+        case MPI_UNSIGNED_LONG:  res = CONDUIT_NATIVE_UNSIGNED_LONG_ID;  break;
+        
+        #if defined(CONDUIT_USE_LONG_LONG)
+        case MPI_UNSIGNED_LONG_LONG:
+        {
+            res = CONDUIT_NATIVE_UNSIGNED_LONG_LONG_ID;
+            break;
+        }
+        #endif
+
+        // floating point types
+
+        case MPI_FLOAT:  res = CONDUIT_NATIVE_FLOAT_ID;  break;
+        case MPI_DOUBLE: res = CONDUIT_NATIVE_DOUBLE_ID; break;
+
+        #if defined(CONDUIT_USE_LONG_DOUBLE)
+        case MPI_LONG_DOUBLE: res = CONDUIT_NATIVE_LONG_DOUBLE_ID; break;
+        #endif
+
+    }
+
+    return res;
+}
+
 
 //---------------------------------------------------------------------------//
 int 
-send(Node &node, int dest, int tag, MPI_Comm comm)
+send_using_schema(const Node &node, int dest, int tag, MPI_Comm comm)
 { 
 
-    Schema schema_c;
-    node.schema().compact_to(schema_c);
-    std::string schema = schema_c.to_json();
-    int schema_len = schema.length() + 1;
-
-    std::vector<uint8> data;
-    node.serialize(data);
-    int data_len = data.size();
+    Node snd_compact;
+    std::string snd_schema = "";
 
+    const void *snd_ptr = node.contiguous_data_ptr();
+    int         snd_data_size = node.total_bytes_compact();
+    
+    if(snd_ptr != NULL && 
+       node.is_compact() )
+    {
+        snd_schema = node.schema().to_json();
+    }
+    else
+    {
+        node.compact_to(snd_compact);
+        snd_schema = snd_compact.schema().to_json();
+        snd_ptr = snd_compact.data_ptr();
+    }
+    
+    int snd_schema_size = snd_schema.length() + 1;
 
-    int intArray[2] = { schema_len, data_len };
 
```

Comment:
On thing that constrains 2 vs 3 messages -- if we use one big-buffer,  we can directly pass the existing buffer to MPI (in the case where that is allowed). Its a chicken and egg issue as well, we can't identify this case until the schema is parsed. 

Can we avoid the first message (with counts) totally by sending the schema, and using mpi probe?